File: NetDeck/IphoneIdentityViewController.m
//
//  IphoneIdentityViewController.m
//  Net Deck
//
//  Created by Gereon Steffens on 18.08.15.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "IphoneIdentityViewController.h"
#import "EditDeckViewController.h"

@interface IphoneIdentityViewController ()

@property Card* selectedIdentity;
@property NSIndexPath* selectedIndexPath;
@property NSArray<NSArray<Card*>*>* identities;
@property NSArray<NSString*>* factionNames;

@end

@implementation IphoneIdentityViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.view.backgroundColor = [UIColor colorWithPatternImage:[ImageCache hexTile]];
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    self.tableView.backgroundColor = [UIColor clearColor];
    
    UITapGestureRecognizer* tableTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(doubleTap:)];
    tableTap.numberOfTapsRequired = 2;
    [self.tableView addGestureRecognizer:tableTap];
    
    self.title = l10n(@"Choose Identity");
    
    
    self.cancelButton.title = l10n(@"Cancel");
    if (!self.deck)
    {
        NSMutableArray* barButtons = self.toolbar.items.mutableCopy;
        
        [barButtons removeObject:self.cancelButton];
        self.toolbar.items = barButtons;
    }
    
    self.selectedIdentity = self.deck.identity;
    
    TableData* identities = [CardManager identitiesForSelection:self.role];
    self.factionNames = identities.sections;
    self.identities = identities.values;
    
    self.selectedIndexPath = nil;
    for (NSInteger i=0; i<self.identities.count; ++i) {
        NSArray<Card*>* arr = self.identities[i];
        for (NSInteger j=0; j<arr.count; ++j) {
            Card* card = arr[j];
            if ([self.selectedIdentity.code isEqual:card.code]) {
                self.selectedIndexPath = [NSIndexPath indexPathForRow:j inSection:i];
                break;
            }
        }
    }
}

-(void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.okButton.enabled = self.deck != nil && self.deck.identity != nil;
}

-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    if (self.selectedIndexPath)
    {
        [self.tableView selectRowAtIndexPath:self.selectedIndexPath animated:NO scrollPosition:UITableViewScrollPositionMiddle];
    }
}


-(void) doubleTap:(UITapGestureRecognizer*)gesture
{
    if (UIGestureRecognizerStateEnded != gesture.state) {
        return;
    }
    
    CGPoint point = [gesture locationInView:self.tableView];
    NSIndexPath* indexPath = [self.tableView indexPathForRowAtPoint:point];
    if (indexPath == nil || self.selectedIdentity == nil) {
        return;
    }
    
    [self okClicked:nil];
}

-(void) okClicked:(id)sender
{
    if (self.deck)
    {
        [self.deck addCard:self.selectedIdentity copies:1];
        
        [self cancelClicked:sender];
    }
    else
    {
        Deck* deck = [[Deck alloc] init];
        deck.role = self.role;
        if (self.selectedIdentity)
        {
            [deck addCard:self.selectedIdentity copies:1];
        }
        
        NSInteger seq = [[NSUserDefaults standardUserDefaults] integerForKey:SettingsKeys.FILE_SEQ] + 1;
        deck.name = [NSString stringWithFormat:@"Deck #%ld", (long)seq];

        EditDeckViewController* edit = [[EditDeckViewController alloc] initWithNibName:@"EditDeckViewController" bundle:nil];
        edit.deck = deck;
        
        NSMutableArray* vcs = self.navigationController.viewControllers.mutableCopy;
        [vcs removeLastObject];
        [vcs addObject:edit];
        [self.navigationController setViewControllers:vcs animated:YES];
    }
}

-(void) cancelClicked:(id)sender
{
    NSAssert(self.deck, @"no deck?");
    
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - tableview

-(NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSArray* arr = self.identities[section];
    return arr.count;
}

-(NSInteger) numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.identities.count;
}

-(NSString*) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return self.factionNames[section];
}

-(void) tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    cell.textLabel.font = [UIFont systemFontOfSize:17];
    cell.backgroundColor = [UIColor whiteColor];
    if ([indexPath compare:self.selectedIndexPath] == NSOrderedSame)
    {
        cell.textLabel.font = [UIFont boldSystemFontOfSize:17];
        cell.backgroundColor = [UIColor colorWithWhite:.97 alpha:1];
    }
}

-(UITableViewCell*) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString* cellIdentifier = @"idCell";
    
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    
    Card* card = [self.identities objectAtIndexPath:indexPath];
    cell.textLabel.text = card.name;
    cell.textLabel.textColor = card.factionColor;
 
    NSString* influence = card.influenceLimit == -1 ? @"∞" : [NSString stringWithFormat:@"%ld", (long)card.influenceLimit];
    if (self.role == NRRoleRunner)
    {
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%ld/%@ · %ld Link", (long)card.minimumDecksize, influence, (long)card.baseLink];
    }
    else
    {
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%ld/%@", (long)card.minimumDecksize, influence];
    }
    
    return cell;
}

-(void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    Card* card = [self.identities objectAtIndexPath:indexPath];
    self.selectedIdentity = card;
    
    NSMutableArray* reload = [NSMutableArray array];
    [reload addObject:indexPath];
    if (self.selectedIndexPath)
    {
        [reload addObject:self.selectedIndexPath];
    }
    self.selectedIndexPath = indexPath;
    [self.tableView reloadRowsAtIndexPaths:reload withRowAnimation:UITableViewRowAnimationNone];
    
    self.okButton.enabled = YES;
}


@end

File: NetDeck/IphoneIdentityViewController.swift
//
//  IphoneIdentityViewController.swift
//  NetDeck
//
//  Created by Gereon Steffens on 31.10.16.
//  Copyright © 2017 Gereon Steffens. All rights reserved.
//

import UIKit
import SwiftyUserDefaults

class IphoneIdentityViewController: UIViewController {

    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var okButton: UIBarButtonItem!
    @IBOutlet weak var cancelButton: UIBarButtonItem!

    @IBOutlet weak var toolbar: UIToolbar!
    
    var role = Role.none
    var deck: Deck?
    
    fileprivate var selectedIdentity: Card?
    fileprivate var selectedIndexPath: IndexPath?
    fileprivate var identities = [[Card]]()
    fileprivate var factionNames = [String]()
    
    override func viewDidLoad() {
        self.view.backgroundColor = UIColor(patternImage: ImageCache.hexTile)
        self.tableView.tableFooterView = UIView(frame: CGRect.zero)
        self.tableView.backgroundColor = .clear
        
        let tableTap = UITapGestureRecognizer(target: self, action: #selector(self.doubleTap(_:)))
        tableTap.numberOfTapsRequired = 2
        self.tableView.addGestureRecognizer(tableTap)
        
        let longPress = UILongPressGestureRecognizer(target: self, action: #selector(self.longPress(_:)))
        self.tableView.addGestureRecognizer(longPress)
        
        self.title = "Choose Identity".localized()
        
        self.cancelButton.title = "Cancel".localized()
        if self.deck == nil {
            if var barButtons = self.toolbar.items, let index = barButtons.index(where: { $0 == self.cancelButton }) {
                barButtons.remove(at: index)
                self.toolbar.items = barButtons
            }
        }
        
        self.selectedIdentity = self.deck?.identity

        let packUsage = Defaults[.deckbuilderPacks]
        let identities = CardManager.identitiesForSelection(self.role, packUsage: packUsage)
        self.factionNames = identities.sections 
        self.identities = identities.values
        
        self.selectedIndexPath = nil
        for i in 0 ..< self.identities.count {
            for j in 0 ..< self.identities[i].count {
                let card = self.identities[i][j]
                if self.selectedIdentity?.code == card.code {
                    self.selectedIndexPath = IndexPath(row: j, section: i)
                    break
                }
            }
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.okButton.isEnabled = self.deck != nil && self.deck?.identity != nil
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        if let selected = self.selectedIndexPath {
            self.tableView.selectRow(at: selected, animated: false, scrollPosition: .middle)
        }
    }

    @IBAction func okTapped(_ sender: UIBarButtonItem) {
        self.closeAndSetIdentity()
    }
    
    func closeAndSetIdentity() {
        guard let identity = self.selectedIdentity else { return }
        if let deck = self.deck {
            deck.addCard(identity, copies: 1)
            self.close()
        } else {
            let deck = Deck(role: self.role)
            deck.addCard(identity, copies: 1)
            
            let edit = EditDeckViewController()
            edit.deck = deck
            
            if var controllers = self.navigationController?.viewControllers {
                controllers.removeLast()
                controllers.append(edit)
                self.navigationController?.setViewControllers(controllers, animated: false)
            }
        }
    }
    
    @IBAction func cancelTapped(_ sender: UIBarButtonItem) {
        assert(self.deck != nil, "no deck")
        self.close()
    }
    
    func close() {
        let _ = self.navigationController?.popViewController(animated: true)
    }
}

// MARK: - tableview delegate & datasource

extension IphoneIdentityViewController: UITableViewDelegate, UITableViewDataSource {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.factionNames.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.identities[section].count
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return self.factionNames[section]
    }
    
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        cell.textLabel?.font = UIFont.systemFont(ofSize: 17)
        cell.backgroundColor = .white
        if indexPath == self.selectedIndexPath {
            cell.textLabel?.font = UIFont.boldSystemFont(ofSize: 17)
            cell.backgroundColor = UIColor(white: 0.97, alpha: 1.0)
        }
    }
 
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let card = self.identities[indexPath.section][indexPath.row]
        self.selectedIdentity = card
        
        var reload = [indexPath]
        if let prev = self.selectedIndexPath {
            reload.append(prev)
        }
        self.selectedIndexPath = indexPath
        tableView.reloadRows(at: reload, with: .none)
        
        self.okButton.isEnabled = true
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let identifier = "idCell"
        
        let cell = tableView.dequeueReusableCell(withIdentifier: identifier) ?? {
            let c = UITableViewCell(style: .subtitle, reuseIdentifier: identifier)
            c.selectionStyle = .none
            return c
        }()
        
        let card = self.identities[indexPath.section][indexPath.row]
        cell.textLabel?.text = card.name
        cell.textLabel?.textColor = card.factionColor
        
        let influence = card.influenceLimit == -1 ? "∞" : "\(card.influenceLimit)"
        if self.role == .runner {
            cell.detailTextLabel?.text = "\(card.minimumDecksize)/\(influence) · \(card.baseLink) Link"
        } else {
            cell.detailTextLabel?.text = "\(card.minimumDecksize)/\(influence)"
        }
        return cell
    }
    
}

// MARK: - gesture recognizers

extension IphoneIdentityViewController {
    
    @objc func doubleTap(_ gesture: UITapGestureRecognizer) {
        guard gesture.state == .ended else { return }
        
        let point = gesture.location(in: self.tableView)
        let indexPath = self.tableView.indexPathForRow(at: point)
        if indexPath == nil || self.selectedIdentity == nil {
            return
        }
        
        self.closeAndSetIdentity()
    }
    
    @objc func longPress(_ gesture: UILongPressGestureRecognizer) {
        guard gesture.state == .began else { return }
        
        let point = gesture.location(in: self.tableView)
        if let indexPath = self.tableView.indexPathForRow(at: point) {
            let card = self.identities[indexPath.section][indexPath.row]
            
            let imageView = ModalCardImageViewController()
            let presenter = ModalViewPresenter(with: imageView)
            
            imageView.modalPresenter = presenter
            imageView.card = card
            presenter.present(on: self)
        }

    }
}